pkg/reconciler/configrevision_reconciler.go

```diff
@@ -0,0 +1,457 @@
+		if cnt, err = crr.processConfigsForRevision(ctx, nodeConfigs.Items, &revisions.Items[i]); err != nil {
+			return fmt.Errorf("failed to process configs for revision %s: %w", revisions.Items[i].Name, err)
+		}
+		cntMap[revisions.Items[i].Spec.Revision] = cnt
+	}
+
+	revisionToDeploy := getFirstValidRevision(revisions.Items)
+
+	nodesToDeploy := getOutdatedNodes(nodes, nodeConfigs.Items, revisionToDeploy)
+
+	if err := crr.updateRevisionCounters(ctx, revisions.Items, revisionToDeploy, len(nodesToDeploy), totalNodes, cntMap); err != nil {
+		return fmt.Errorf("failed to update queue counters: %w", err)
+	}
+
+	// there is nothing to deploy - skip
+	if revisionToDeploy == nil {
+		crr.logger.Error(fmt.Errorf("there is no revision to deploy"), "revision deployment aboorted")
+		return nil
+	}
+
+	if revisionToDeploy.Status.Ongoing < crr.maxUpdating && len(nodesToDeploy) > 0 {
+		if err := crr.deployNodeConfig(ctx, nodesToDeploy[0], revisionToDeploy); err != nil {
+			return fmt.Errorf("error deploying node configurations: %w", err)
+		}
+	}
+
+	// remove all but last known valid revision
+	if err := crr.revisionCleanup(ctx); err != nil {
+		return fmt.Errorf("error cleaning redundant revisions: %w", err)
+	}
+
+	return nil
+}
+
+func getFirstValidRevision(revisions []v1alpha1.NetworkConfigRevision) *v1alpha1.NetworkConfigRevision {
+	i := slices.IndexFunc(revisions, func(r v1alpha1.NetworkConfigRevision) bool {
+		return !r.Status.IsInvalid
+	})
+	if i > -1 {
+		return &revisions[i]
+	}
+	return nil
+}
+
+type counters struct {
+	ready, ongoing, invalid int
+}
+
+func (crr *ConfigRevisionReconciler) processConfigsForRevision(ctx context.Context, configs []v1alpha1.NodeNetworkConfig, revision *v1alpha1.NetworkConfigRevision) (*counters, error) {
+	configs, err := crr.removeRedundantConfigs(ctx, configs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to remove redundant configs: %w", err)
+	}
+	ready, ongoing, invalid := getRevisionCounters(configs, revision)
+	cnt := &counters{ready: ready, ongoing: ongoing, invalid: invalid}
+
+	if invalid > 0 {
+		if err := crr.invalidateRevision(ctx, revision, "NetworkConfigRevision results in invalid config"); err != nil {
+			return cnt, fmt.Errorf("faild to invalidate revision %s: %w", revision.Name, err)
+		}
+	}
+
+	return cnt, nil
+}
+
+func getRevisionCounters(configs []v1alpha1.NodeNetworkConfig, revision *v1alpha1.NetworkConfigRevision) (ready, ongoing, invalid int) {
+	ready = 0
+	ongoing = 0
+	invalid = 0
+	for i := range configs {
+		if configs[i].Spec.Revision == revision.Spec.Revision {
+			switch configs[i].Status.ConfigStatus {
+			case StatusInvalid:
+				// Increase 'invalid' counter so we know that there the revision results in invalid configs.
+				invalid++
+			case StatusProvisioning, "":
+				// Update ongoing counter
+				ongoing++
+				if wasConfigTimeoutReached(&configs[i]) {
+					// If timout was reached revision is invalid (but still counts as ongoing).
+					invalid++
+				}
```

Comment:
Okay, I am struggling with this a bit. 

Imagine the following scenario:
1. A new node joins the cluster (or network operator is updated etc)
2. network-operator is not running for various reasons for 2 minutes

I am thinking of a better solution to find out if a rollout has reached a timeout.

Some things to consider:
- If a node can not rollout config at all because of *other issues* the node should be marked as NotReady by the kubelet anyway, our automation will take care of removing that node alltogether (deleting the nodeconfig anyway)
- A rollout can only timeout once it was started

What would you think of a `applyStarted` field in status which is written **before** a nodeconfig is rolled out (aka an agent has received the config but has not yet configured it). This is used to also update the lastUpdated timestamp and if applyStarted is true and lastUpdated has reached the timeout of 2 minutes the config is marked as failed.

Reply:
Actually we already have `provisioning`, maybe just use that instead of `applyStarted`

Reply:
but wait a second, that's already the case, let me look further..

Reply:
What if, before invalidating the revision, we would check if node that 'timed out' is still available (e.g it was not deleted or not in `NotReady` state)? And if it was deleted or NotReady, we could then simply not use it's config's state to determine if revision should be invalidated.

Reply:
>     * A rollout can only timeout once it was started

Currently that's true, as only the agent that operates on the node updates the `LastUpdate` timestamp in the config's status, so if there's any value there, it means that node stared to reconfigure itself.

Reply:
I see one issue: The ConfigStatus is by default an empty string and defaults to "", the lastUpdate is 0 (aka 1990...) if we check for "" as well in line 164: `case StatusProvisioning, "":` and then check for the timeout it will always be true I guess?

ConfigStatus == "" should be a different case with a higher timeout first of all **plus** a check that lastUpdate is not 0

Reply:
Also, made timeout value configurable using flags.